test_container/tests/test/pandas/all/emit_dtypes.py

```diff
@@ -0,0 +1,369 @@
+            ("boolean", "boolean", nan_dataframe_value_str, nan_expected_rows, False),
+            ("object", "boolean", nan_dataframe_value_str, ".*F-UDF-CL-SL-PYTHON-1068: emit column 0 of type BOOLEAN but data given have type py_float.*", False),
+
+            ("datetime64[ns]", "timestamp", nan_dataframe_value_str, nan_expected_rows, False),
+            ("object", "timestamp", nan_dataframe_value_str, ".*F-UDF-CL-SL-PYTHON-1068: emit column 0 of type TIMESTAMP but data given have type py_float.*", False),
+            ("object", "DATE", nan_dataframe_value_str, ".*F-UDF-CL-SL-PYTHON-1068: emit column 0 of type DATE but data given have type py_float.*", False),
+
+        ]
+
+    @useData(types)
+    def test_dtype_emit(self, dtype:str, sql_type:str, dataframe_value_str:str, expected_result:Union[str,List[Tuple]], use_almost_equal:bool):
+        sql=udf.fixindent(f'''
+            CREATE OR REPLACE PYTHON3 SET SCRIPT test_dtype_emit(i integer) 
+            EMITS (o1 {sql_type}, o2 {sql_type}, traceback varchar(2000000)) AS
+
+            def run(ctx):
+                try:
+                    from decimal import Decimal
+                    import pandas as pd
+                    import numpy as np
+                    from datetime import datetime, date
+                    df = pd.DataFrame({dataframe_value_str}, dtype="{dtype}")
+                    df["traceback"]=None
+                    ctx.emit(df)
+                except:
+                    import traceback
+                    ctx.emit(None,None,traceback.format_exc())
+            /
+            ''')
+        print(sql)
+        self.query(sql)
+        with UdfDebugger(test_case=self):
+            rows = self.query('''SELECT test_dtype_emit(0)''')
+            if isinstance(expected_result,str):
+                self.assertRegex(rows[0][2], expected_result)
+            else:
+                if use_almost_equal:
+                    self.assertRowsAlmostEqual(expected_result, rows, places=1)
+                else:
+                    self.assertRowsEqual(expected_result, rows)
+
+    def isValueAlmostEqual(self, left, right, places):
+        if isinstance(left, (float, Decimal)) and isinstance(right, (float, Decimal)):
+            return round(left, places) == round(right, places)
+        else:
+            return left == right
+
+    def isRowAlmostEqual(self, left, right, places):
+        if len(left) != len(right):
+            return False
+        all_values_almost_equal = all(self.isValueAlmostEqual(lvalue, rvalue, places)
+                                      for lvalue, rvalue in zip(left, right))
+        return all_values_almost_equal
+ 
+    def assertRowsAlmostEqual(self, left, right, places):
+        lrows = [tuple(x) for x in left]
+        rrows = [tuple(x) for x in right]
+        if len(lrows) != len(rrows):
+            raise AssertionError(f'{lrows} and {rrows} have different number of rows.')
```

Comment:
why not simply:
```suggestion
        assert len(lrows) == len(rrows)
```

Reply:
this is not pytest, assert doesn't work here so well